README.md
Healthcheck-Extended
----------

Healthcheck wraps a Flask app object and adds a way to write simple healthcheck
functions that can be used to monitor your application. It's useful for
asserting that your dependencies are up and running and your application can
respond to HTTP requests. The Healthcheck functions are exposed via a user
defined flask route so you can use an external monitoring application (monit,
nagios, Runscope, etc.) to check the status and uptime of your application.

New in version 1.1: Healthcheck also gives you a simple Flask route to view
information about your application's environment. By default, this includes
data about the operating system, the Python environment, the current process,
and the application config. You can customize which sections are included, or
add your own sections to the output.

This fork contains some tweaks, mostly regarding a more flexible timestamp
format handling to make it more usable.

Forked from: https://github.com/Runscope/healthcheck

### Version 1.3.4
- Add support for UTC timestamps

## Installing

```
pip install healthcheck-extended

```

## Usage

Here's an example of basic usage:

```python
from flask import Flask
from healthcheck_extended import HealthCheck, EnvironmentDump

app = Flask(__name__)

# wrap the flask app and give a heathcheck url
health = HealthCheck(app, "/healthcheck")
envdump = EnvironmentDump(app, "/environment")

# add your own check function to the healthcheck
def redis_available():
    client = _redis_client()
    info = client.info()
    return True, "redis ok"

health.add_check(redis_available)

# add your own data to the environment dump
def application_data():
	return {"maintainer": "Frank Stratton",
	        "git_repo": "https://github.com/Runscope/healthcheck"}

envdump.add_section("application", application_data)
```

To run all of your check functions, make a request to the healthcheck URL
you specified, like this:

```
curl "http://localhost:5000/healthcheck"
```

And to view the environment data, make a check to the URL you specified for
EnvironmentDump:

```
curl "http://localhost:5000/environment"
```

## The HealthCheck class

### Check Functions

Check functions take no arguments and should return a tuple of (bool, str).
The boolean is whether or not the check passed. The message is any string or
output that should be rendered for this check. Useful for error
messages/debugging.

```python
# add check functions
def addition_works():

	if 1 + 1 == 2:
		return True, "addition works"
	else:
		return False, "the universe is broken"
```

Any exceptions that get thrown by your code will be caught and handled as
errors in the healthcheck:

```python
# add check functions
def throws_exception():
	bad_var = None
	bad_var['explode']

```

Will output:

```json
{
	"status": "failure",
		"results": [
		{
			"output": "'NoneType' object has no attribute '__getitem__'",
			"checker": "throws_exception",
			"passed": false
		}
	]
}
```

Note, all checkers will get run and all failures will be reported. It's
intended that they are all separate checks and if any one fails the
healthcheck overall is failed.

### Caching

In Runscope's infrastructure, the /healthcheck endpoint is hit surprisingly
often. haproxy runs on every server, and each haproxy hits every healthcheck
twice a minute. (So if we have 30 servers in our infrastructure, that's 60
healthchecks per minute to every Flask service.) Plus, monit hits every
healthcheck 6 times a minute. 

To avoid putting too much strain on backend services, health check results can
be cached in process memory. By default, health checks that succeed are cached
for 27 seconds, and failures are cached for 9 seconds. These can be overridden
with the `success_ttl` and `failed_ttl` parameters. If you don't want to use
the cache at all, initialize the Healthcheck object with `success_ttl=None,
failed_ttl=None`.

### Customizing

You can customize the status codes, headers, and output format for success and
failure responses.

## The EnvironmentDump class

### Built-in data sections

By default, EnvironmentDump data includes these 4 sections:

* `os`: information about your operating system.
* `python`: information about your Python executable, Python path, and
installed packages.
* `process`: information about the currently running Python process, including
the PID, command line arguments, and all environment variables.
* `config`: information about your Flask app's configuration, pulled from
`app.config`.

Some of the data is scrubbed to avoid accidentally exposing passwords or access
keys/tokens. Config keys and environment variable names are scanned for `key`,
`token`, or `pass`. If those strings are present in the name of the variable,
the value is not included.

### Disabling built-in data sections

For security reasons, you may want to disable an entire section. You can
disable sections when you instantiate the `EnvironmentDump` object, like this:

```python
envdump = EnvironmentDump(app, "/environment",
                          include_python=False, include_os=False,
                          include_process=False, include_config=False)
```

### Adding custom data sections

You can add a new section to the output by registering a function of your own.
Here's an example of how this would be used:

```python
def application_data():
	return {"maintainer": "Frank Stratton",
	        "git_repo": "https://github.com/Runscope/healthcheck"}

envdump = EnvironmentDump(app, "/environment")
envdump.add_section("application", application_data)
```


CHANGELOG.md
# Healthcheck Changelog

### 1.3.4
- Add support for UTC timestamps

### 1.3.1
- Fix for 'Inappropriate ioctl for device' error on posix systems.

### 1.3.0
- Adds support for init_app construction of healthcheck objects. Thanks to
  Iuri de Silvio for the pull request.


### 1.2.0
- Adds support for Python 3.x. Thanks to Guilherme D'Amoreira for the pull
request.

### 1.1.0
- Added the `EnvironmentDump` class which provides a second endpoint for
details about your application's environment.

### 1.0.0
- Incremented the version number to indicate that this is a stable release.

### 0.2
- Added caching to health check responses. Successful checks are cached for 27
seconds; failures are cached for 9 seconds.
- Removed the "simple" view of the health check, which had been available with
the query string `?simple=true`. This isn't necessary now that we cache the
results of the checks.
- Added `timestamp` field to the outer level of the JSON response, and
`timestamp` and `expires` to each check result.


tests/unit/test_healthcheck.py
from __future__ import with_statement

import unittest
import flask
from healthcheck_extended import HealthCheck, EnvironmentDump


class BasicHealthCheckTest(unittest.TestCase):

    def setUp(self):
        self.path = '/h'
        self.app = flask.Flask(__name__)
        self.hc = self._hc()
        self.client = self.app.test_client()

    def _hc(self):
        return HealthCheck(self.app, self.path)

    def test_basic_check(self):
        response = self.client.get(self.path)
        self.assertEqual(200, response.status_code)

    def test_failing_check(self):
        def fail_check():
            return False, "FAIL"

        self.hc.add_check(fail_check)
        response = self.client.get(self.path)
        self.assertEqual(500, response.status_code)


class BasicEnvironmentDumpTest(unittest.TestCase):

    def setUp(self):
        self.path = '/e'
        self.app = flask.Flask(__name__)
        self.hc = self._hc()
        self.client = self.app.test_client()

    def _hc(self):
        return EnvironmentDump(self.app, self.path)

    def test_basic_check(self):
        def test_ok():
            return "OK"

        self.hc.add_section("test_func", test_ok)

        response = self.client.get(self.path)
        self.assertEqual(200, response.status_code)
        jr = flask.json.loads(response.data)
        self.assertEqual("OK", jr["test_func"])


class LazyHealthCheckTest(BasicHealthCheckTest):

    def setUp(self):
        super(LazyHealthCheckTest, self).setUp()
        self.hc.init_app(self.app, self.path)

    def _hc(self):
        return HealthCheck()


class LazyEnvironmentDumpTest(unittest.TestCase):

    def setUp(self):
        super(LazyEnvironmentDumpTest, self).setUp()
        self.hc.init_app(self.app, self.path)

    def _hc(self):
        return EnvironmentDump()


if __name__ == '__main__':
    unittest.main()


healthcheck/__init__.py
import imp
import json
import os
import six
import socket
import sys
import time
import traceback
import platform
import datetime
from flask import current_app
try:
    from functools import reduce
except Exception:
    pass


def basic_exception_handler(_, e):
    return False, str(e)


def json_success_handler(results):
    timestamp = Timestamp(time_format="utcnow")
    data = {
        'hostname': socket.gethostname(),
        'status': 'success',
        'timestamp': str(timestamp.get_time()),
        'results': results,
    }

    return json.dumps(data)


def json_failed_handler(results):
    timestamp = Timestamp(time_format="utcnow")
    data = {
        'hostname': socket.gethostname(),
        'status': 'failure',
        'timestamp': str(timestamp.get_time()),
        'results': results,
    }

    return json.dumps(data)


def check_reduce(passed, result):
    return passed and result.get('passed')


class HealthCheck(object):
    def __init__(self, app=None, path=None, success_status=200,
                 success_headers=None, success_handler=json_success_handler,
                 success_ttl=27, failed_status=500, failed_headers=None,
                 failed_handler=json_failed_handler, failed_ttl=9,
                 exception_handler=basic_exception_handler, checkers=None,
                 log_on_failure=True, time_format="time",
                 **options):
        self.cache = dict()

        self.success_status = success_status
        self.success_headers = success_headers or {'Content-Type': 'application/json'}
        self.success_handler = success_handler
        self.success_ttl = float(success_ttl or 0)

        self.failed_status = failed_status
        self.failed_headers = failed_headers or {'Content-Type': 'application/json'}
        self.failed_handler = failed_handler
        self.failed_ttl = float(failed_ttl or 0)

        self.exception_handler = exception_handler

        self.log_on_failure = log_on_failure

        self.options = options
        self.checkers = checkers or []

        self.timestamp = Timestamp(time_format="utcnow")

        if app:
            self.init_app(app, path)

    def init_app(self, app, path):
        if path:
            app.add_url_rule(path, view_func=self.check, **self.options)

    def add_check(self, func):
        self.checkers.append(func)

    def check(self):
        results = []
        for checker in self.checkers:
            timestamp = Timestamp(time_format="utcnow")
            if checker in self.cache and self.cache[checker].get('expires') >= str(timestamp.get_time()):
                result = self.cache[checker]
            else:
                result = self.run_check(checker)
                self.cache[checker] = result
            results.append(result)

        passed = reduce(check_reduce, results, True)

        if passed:
            message = "OK"
            if self.success_handler:
                message = self.success_handler(results)

            return message, self.success_status, self.success_headers
        else:
            message = "NOT OK"
            if self.failed_handler:
                message = self.failed_handler(results)

            return message, self.failed_status, self.failed_headers

    def run_check(self, checker):
        try:
            passed, output = checker()
        except Exception:
            traceback.print_exc()
            e = sys.exc_info()[0]
            current_app.logger.exception(e)
            passed, output = self.exception_handler(checker, e)

        if not passed:
            msg = 'Health check "{}" failed with output "{}"'.format(checker.__name__, output)
            if self.log_on_failure:
                current_app.logger.error(msg)

        timestamp = self.timestamp.get_time()
        if passed:
            expires = timestamp + datetime.timedelta(seconds=self.success_ttl)
        else:
            expires = timestamp + datetime.timedelta(seconds=self.failed_ttl)

        result = {'checker': checker.__name__,
                  'output': output,
                  'passed': passed,
                  'timestamp': str(timestamp),
                  'expires': str(expires)}
        return result


class Timestamp(object):
    def __init__(self, time_format="time"):
        self.time_format = time_format
        self._set_time_format()

    def _set_time_format(self):
        if self.time_format == "time":
            self.time_function = datetime.datetime.time
        elif self.time_format == "utcnow":
            self.time_function = datetime.datetime.utcnow

    def get_time(self):
        return self.time_function()


class EnvironmentDump(object):
    def __init__(self, app=None, path=None,
                 include_os=True, include_python=True,
                 include_config=True, include_process=True):
        self.functions = {}
        if include_os:
            self.functions['os'] = self.get_os
        if include_python:
            self.functions['python'] = self.get_python
        if include_config:
            self.functions['config'] = self.get_config
        if include_process:
            self.functions['process'] = self.get_process

        if app:
            self.init_app(app, path)

    def init_app(self, app, path):
        if path:
            app.add_url_rule(path, view_func=self.dump_environment)

    def add_section(self, name, func):
        if name in self.functions:
            raise Exception('The name "{}" is already taken.'.format(name))
        self.functions[name] = func

    def dump_environment(self):
        data = {}
        for (name, func) in six.iteritems(self.functions):
            data[name] = func()

        return json.dumps(data), 200, {'Content-Type': 'application/json'}

    def get_os(self):
        return {'platform': sys.platform,
                'name': os.name,
                'uname': platform.uname()}

    def get_config(self):
        return self.safe_dump(current_app.config)

    def get_python(self):
        result = {'version': sys.version,
                  'executable': sys.executable,
                  'pythonpath': sys.path,
                  'version_info': {'major': sys.version_info.major,
                                   'minor': sys.version_info.minor,
                                   'micro': sys.version_info.micro,
                                   'releaselevel': sys.version_info.releaselevel,
                                   'serial': sys.version_info.serial}}
        if imp.find_module('pkg_resources'):
            import pkg_resources
            packages = dict([(p.project_name, p.version) for p in pkg_resources.working_set])
            result['packages'] = packages

        return result

    def get_login(self):
        # Based on https://github.com/gitpython-developers/GitPython/pull/43/
        # Fix for 'Inappopropirate ioctl for device' on posix systems.
        if os.name == "posix":
            import pwd
            username = pwd.getpwuid(os.geteuid()).pw_name
        else:
            username = os.environ.get('USER', os.environ.get('USERNAME', 'UNKNOWN'))
            if username == 'UNKNOWN' and hasattr(os, 'getlogin'):
                username = os.getlogin()
        return username

    def get_process(self):
        return {'argv': sys.argv,
                'cwd': os.getcwd(),
                'user': self.get_login(),
                'pid': os.getpid(),
                'environ': self.safe_dump(os.environ)}

    def safe_dump(self, dictionary):
        result = {}
        for key in dictionary.keys():
            if 'key' in key.lower() or 'token' in key.lower() or 'pass' in key.lower():
                # Try to avoid listing passwords and access tokens or keys in the output
                result[key] = "********"
            else:
                try:
                    json.dumps(dictionary[key])
                    result[key] = dictionary[key]
                except TypeError:
                    pass
        return result
